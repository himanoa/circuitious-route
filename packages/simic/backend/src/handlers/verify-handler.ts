import * as Express from "express";
import { SQL, SQLStatement } from "sql-template-strings";
import { TokenNotFoundError } from "../error";

export const verifyHandler = (deps: {
  executeQuery: Promise<(sql: string | SQLStatement) => Promise<any>>;
  generateRandomString: () => string;
  verify: (token: string) => { discordId: string };
}) => async (req: Express.Request, res: Express.Response) => {
  try {
    const executeQuery = await deps.executeQuery;
    const tokenWithType = req.get("Authorization");
    if (!tokenWithType) {
      throw new TokenNotFoundError("Token is not found");
    }
    const [, token] = tokenWithType.split(" ");
    const { discordId } = deps.verify(token);
    const profiles = await executeQuery(
      SQL`SELECT * FROM profiles WHERE discord_id = ${discordId};`
    );
    res.status(200).json({
      profiles: profiles.map(
        (p: { stream_key: string; discord_id: string }) => {
          return {
            streamKey: p.stream_key,
            discordId: p.discord_id,
          };
        }
      ),
      discordId,
    });
  } catch (err) {
    res.status(401).json({ error: err });
  }
};
