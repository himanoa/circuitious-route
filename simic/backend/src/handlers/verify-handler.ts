import  * as Express from "express";
import { SQL, SQLStatement } from "sql-template-strings"
import { TokenNotFoundError} from "../error"

export const verifyHandler = (
  deps: {
    executeQuery: Promise<(sql: string | SQLStatement) => Promise<any>>,
    generateRandomString: () => string,
    verify: (token: string) => {discordId: string}
  }
) => async (req: Express.Request, res: Express.Response) => {
  try {
    const executeQuery = await deps.executeQuery
    const tokenWithType = req.get("Authorization")
    if(!tokenWithType) {
      throw new TokenNotFoundError("Token is not found")
    }
    const [,token] = tokenWithType.split(" ")
    const { discordId } = deps.verify(token)
    const profiles = await executeQuery(SQL`SELECT * FROM user WHERE profiles.discord_id = ${parseInt(discordId, 10)}`)
    res.status(200).json({
      profiles,
      discordId
    })
  } catch {
    res.status(401).json({})
  }
}
