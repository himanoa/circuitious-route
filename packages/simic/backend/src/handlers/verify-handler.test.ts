import { verifyHandler } from "./verify-handler";
import { SQLStatement } from "sql-template-strings";
import * as Response from "jest-express/lib/response";
import * as Request from "jest-express/lib/request";

describe("verifyHandler", () => {
  describe("when verify successful", () => {
    const executedQuery: string[] = [];
    const target = verifyHandler({
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if (typeof sql === "string") {
          executedQuery.push(sql);
          return Promise.resolve([1]);
        }
        executedQuery.push(sql.sql);
        return Promise.resolve([1]);
      }),
      generateRandomString: () => "foobar",
      verify: () => ({ discordId: "12123123".toString() }),
    });

    it("should be failed", async () => {
      const request = new Request.Request("https://example.com");
      request.setHeaders("Authorization", "Barer InvalidToken");
      const response = new Response.Response();
      await target(request as any, response as any);
      expect(executedQuery).toMatchSnapshot();
      expect(response.status).toBeCalledWith(200);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("when verify failed", () => {
    const executedQuery: string[] = [];
    const target = verifyHandler({
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if (typeof sql === "string") {
          executedQuery.push(sql);
          return Promise.resolve([1]);
        }
        executedQuery.push(sql.sql);
        return Promise.resolve([1]);
      }),
      generateRandomString: () => "foobar",
      verify: () => {
        throw new Error("doo");
      },
    });

    it("should be failed", async () => {
      const request = new Request.Request("https://example.com");
      request.setHeaders("Authorization", "Barer InvalidToken");
      const response = new Response.Response();
      await target(request as any, response as any);
      expect(executedQuery).toMatchSnapshot();
      expect(response.status).toBeCalledWith(401);
      expect(response.body).toMatchSnapshot();
    });
  });
});
