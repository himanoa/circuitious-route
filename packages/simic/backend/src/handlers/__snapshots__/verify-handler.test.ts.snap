// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyHandler when verify failed should be failed 1`] = `Array []`;

exports[`verifyHandler when verify failed should be failed 2`] = `
Object {
  "error": [Error: doo],
}
`;

exports[`verifyHandler when verify successful should be failed 1`] = `
Array [
  "SELECT * FROM profiles WHERE discord_id = ?;",
]
`;

exports[`verifyHandler when verify successful should be failed 2`] = `
Object {
  "discordId": "12123123",
  "profiles": Array [
    Object {
      "discordId": undefined,
      "streamKey": undefined,
    },
  ],
}
`;
